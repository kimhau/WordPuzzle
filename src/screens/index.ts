export { default as Congratz } from './Congratz/Congratz';
export { default as Game } from './Game/Game';
export { default as Home } from './Home/Home';
export { default as Leaderboard } from './Leaderboard/Leaderboard';
export { default as Startup } from './Startup/Startup';
