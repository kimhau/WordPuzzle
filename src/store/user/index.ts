import { createSlice } from '@reduxjs/toolkit';

const slice = createSlice({
  name: 'user',
  initialState: { name: null, points: 0 } as UserState,
  reducers: {
    updateUser: (state, { payload: { name, points } }: UserPayload) => {
      if (typeof name !== 'undefined') {
        state.name = name;
      }
      if (typeof points !== 'undefined') {
        state.points = points;
      }
    },
    updatePoints: (state, { payload: { points } }: UserPayload) => {
      if (typeof points !== 'undefined') {
        state.points += points;
      }
    },
  },
});

export const { updateUser, updatePoints } = slice.actions;

export default slice.reducer;

export type UserState = {
  name: string | null;
  points: number;
};

export type UserPayload = {
  payload: Partial<UserState>;
};
