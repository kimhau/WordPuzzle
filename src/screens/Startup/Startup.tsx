import React, { useEffect } from 'react';
import { useTranslation } from 'react-i18next';
import { ActivityIndicator, Text, View } from 'react-native';
import { ApplicationScreenProps } from '../../../@types/navigation';
import { useTheme } from '../../hooks';
import { setDefaultTheme } from '../../store/theme';

const Startup = ({ navigation }: ApplicationScreenProps) => {
  const { Layout, Gutters, Fonts } = useTheme();

  const { t } = useTranslation();

  const init = async () => {
    await new Promise(resolve =>
      setTimeout(() => {
        resolve(true);
      }, 2000),
    );
    await setDefaultTheme({ theme: 'default', darkMode: null });
    navigation.reset({
      index: 0,
      routes: [{ name: 'Home' }],
    });
  };

  useEffect(() => {
    init();
  }, []);

  return (
    <View style={[Layout.fill, Layout.colCenter]}>
      <ActivityIndicator size={'large'} style={[Gutters.largeVMargin]} />
      <Text style={Fonts.textCenter}>{t('welcome:title')}</Text>
    </View>
  );
};

export default Startup;
