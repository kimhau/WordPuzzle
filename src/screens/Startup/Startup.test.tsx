import { act, render } from '@testing-library/react-native';
import React from 'react';
import { Provider } from 'react-redux';
import { store } from 'WordPuzzle/src/store';
import Startup from './Startup';

// jest.mock('../../store/theme');

describe('Startup', () => {
  beforeEach(() => {
    jest.useFakeTimers(); // Mock timers to avoid waiting for 2 seconds
  });

  afterEach(() => {
    jest.useRealTimers(); // Restore real timers
    jest.resetAllMocks(); // Reset all mocks
  });

  it('should render ActivityIndicator and Text components', async () => {
    const { getByTestId, toJSON } = render(
      <Provider store={store}>
        <Startup />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
    expect(getByTestId('activity-indicator')).toBeDefined();
    expect(getByTestId('text')).toBeDefined();
  });

  it('should call init function and navigate to Home screen', async () => {
    const resetMock = jest.fn();
    const navigation = { reset: resetMock };
    const { getByTestId } = render(
      <Provider store={store}>
        <Startup navigation={navigation} />
      </Provider>,
    );
    await act(async () => {
      jest.runAllTimers(); // Advance timers to resolve setTimeout in init function
    });
    expect(resetMock).toHaveBeenCalledWith({
      index: 0,
      routes: [{ name: 'Home' }],
    });
  });

  it('should display title using translation hook', async () => {
    const { getByText } = render(
      <Provider store={store}>
        <Startup />
      </Provider>,
    );
    expect(getByText('welcome:title')).toBeDefined();
  });
});
