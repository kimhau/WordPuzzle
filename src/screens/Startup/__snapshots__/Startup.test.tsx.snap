// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Startup should render ActivityIndicator and Text components 1`] = `
<View
  style={
    [
      {
        "flex": 1,
      },
      {
        "alignItems": "center",
        "flexDirection": "column",
        "justifyContent": "center",
      },
    ]
  }
>
  <ActivityIndicator
    size="large"
    style={
      [
        {
          "marginVertical": 60,
        },
      ]
    }
    testID="activity-indicator"
  />
  <Text
    style={
      {
        "textAlign": "center",
      }
    }
    testID="text"
  >
    welcome:title
  </Text>
</View>
`;
